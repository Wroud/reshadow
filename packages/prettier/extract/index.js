/**
 * This is a trick to extract the prettier internal plugins
 */

const prettier = require('prettier');

const {_parsers, _plugins} = require('./extract-plugin');

prettier.format(`;`, {
    parser: 'css',
    plugins: [require.resolve('./extract-plugin')],
});

prettier.format(`;`, {
    parser: 'babel',
    plugins: [require.resolve('./extract-plugin')],
});

module.exports = {
    _parsers,
    _plugins,
};
