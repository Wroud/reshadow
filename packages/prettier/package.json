{
    "name": "@reshadow/prettier",
    "version": "0.0.1-alpha.11",
    "description": "reshadow prettier plugin",
    "main": "index.js",
    "repository": {
        "type": "git",
        "url": "git+https://github.com/lttb/reshadow.git"
    },
    "author": "Kenzhaev Artur <kenzhaev.artur@gmail.com>",
    "license": "MIT",
    "bugs": {
        "url": "https://github.com/lttb/reshadow/issues"
    },
    "homepage": "https://reshadow.dev",
    "publishConfig": {
        "access": "public"
    },
    "scripts": {
        "prepublish": "../../build.sh"
    },
    "dependencies": {
        "prettier": "1.17.0"
    },
    "devDependencies": {
        "common-tags": "1.8.0"
    }
}
