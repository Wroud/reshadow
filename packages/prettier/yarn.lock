# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


common-tags@1.8.0:
  version "1.8.0"
  resolved "https://registry.yarnpkg.com/common-tags/-/common-tags-1.8.0.tgz#8e3153e542d4a39e9b10554434afaaf98956a937"
  integrity sha512-6P6g0uetGpW/sdyUy/iQQCbFF0kWVMSIVSyYz7Zgjcgh8mgw8PQzDNZeyZ5DQ2gM7LBoZPHmnjz8rUthkBG5tw==

prettier@1.18.2:
  version "1.18.2"
  resolved "https://registry.yarnpkg.com/prettier/-/prettier-1.18.2.tgz#6823e7c5900017b4bd3acf46fe9ac4b4d7bda9ea"
  integrity sha512-OeHeMc0JhFE9idD4ZdtNibzY0+TPHSpSSb9h8FqtP+YnoZZ1sl8Vc9b1sasjfymH3SonAF4QcA2+mzHPhMvIiw==
