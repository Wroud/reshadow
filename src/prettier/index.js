const {_parsers, _plugins} = require('./extract');

const extend = (name, cb) => ({
    ..._plugins[name].printer,
    ...cb(_plugins[name].printer),
});

const isStyledTag = tag => tag.type === 'Identifier' && tag.name === 'styled';
const isStyledCall = tag =>
    tag.type === 'CallExpression' &&
    tag.callee.type === 'Identifier' &&
    tag.callee.name === 'styled';

const isStyledCallee = callee =>
    callee.type === 'TaggedTemplateExpression' &&
    (isStyledTag(callee.tag) || isStyledCall(callee.tag));

const isStyledBody = node =>
    node.type === 'ArrowFunctionExpression' &&
    node.body.type === 'CallExpression' &&
    isStyledCallee(node.body.callee);

module.exports = {
    parsers: {
        css: {
            ..._parsers.css,
            astFormat: 'reshadow-css',
        },
        babel: {
            ..._parsers.babel,
            astFormat: 'reshadow-babel',
        },
    },
    printers: {
        'reshadow-css': extend('css', printer => ({
            print: (...args) => {
                const [path] = args;
                const node = path.getNode();
                const result = printer.print(...args);

                if (node.type === 'selector-tag') {
                    /**
                     * keep the original tag name
                     * @see https://github.com/prettier/prettier/blob/0d0e88cfdf886f8915d25c64391f27f1a931e32e/src/language-css/printer-postcss.js#L344
                     */
                    result.parts[1] = node.value;
                }

                return result;
            },
        })),
        'reshadow-babel': extend('babel', printer => ({
            print: (...args) => {
                const [path] = args;
                const node = path.getNode();

                const result = printer.print(...args);

                if (!isStyledBody(node)) return result;

                if (!result.parts) return result;

                const [functionGroup] = result.parts;

                // Function presented with a pair of an arrow ('=>') and the body
                let [, bodyGroup] = functionGroup.contents.parts;
                const body = bodyGroup.contents;

                // remove indent
                if (body.parts[0].type === 'indent') {
                    body.parts[0] = body.parts[0].contents;
                }

                // replace newline with whitespace
                body.parts[0].parts[0] = ' ';

                /**
                 * remove unnecessary last comma or newline
                 * we should remove it only for 2 elements
                 * for example:
                 */
                // const Button = children.map(x => styled`
                //     div {
                //         color: green;
                //     }
                // `(
                //     <div {...x} size="m">
                //         {x}
                //     </div>,
                // ),
                // );
                if (body.parts.length === 2) {
                    body.parts.splice(-1);
                }

                return result;
            },
            embed: (...args) => {
                const [path] = args;
                const node = path.getNode();

                let parent;
                let isStyled = false;

                if (node.type === 'TemplateLiteral') {
                    parent = path.getParentNode();

                    /**
                     * A hack to force `embed` function act like it is styled component
                     * @see https://github.com/prettier/prettier/blob/0d0e88cfdf886f8915d25c64391f27f1a931e32e/src/language-js/embed.js#L413
                     */
                    if (parent && parent.type === 'TaggedTemplateExpression') {
                        const {tag} = parent;

                        if (
                            tag.type === 'Identifier' &&
                            tag.name === 'styled'
                        ) {
                            isStyled = true;
                            parent.tag.name = 'css';
                        }
                    }
                }

                const result = printer.embed(...args);

                // restore the original name
                if (isStyled) {
                    parent.tag.name = 'styled';
                }

                return result;
            },
        })),
    },
};
