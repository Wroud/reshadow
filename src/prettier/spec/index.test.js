import prettier from 'prettier';
import {stripIndent} from 'common-tags';

const transform = options => code =>
    prettier.format(stripIndent(code), {
        ...options,
        plugins: [require.resolve('..')],
    });

transform.with = options => code => transform(code, options);

describe('prettier', () => {
    describe('css', () => {
        it('should keep the capitalized tag', () => {
            const code = transform({parser: 'css'})`
                Button {
                    color: red;
                }

                Button[disabled] {
                    color: red;
                }

                Container Button content {
                    font-size: 10px;
                }
            `;

            expect(code).toMatchSnapshot();
        });
    });

    describe('js', () => {
        it('should keep the `styled` tag in arrow function and transform the css', () => {
            const code = transform({parser: 'babel'})`
                const Button = ({children, ...props}) =>
                styled\`Box {margin-top: 10px;}
                \`(
                <Box {...props} size="s">{children}</Box>,
                );
            `;

            expect(code).toMatchSnapshot();
        });

        it('should keep the `styled` tag with composition', () => {
            const code = transform({parser: 'babel'})`
                const Button = ({children, ...props}) =>
                styled(style1, style2)\`Box {margin-top: 10px;} button[disabled] {color: gray;}
                \`(
                <Box {...props} size="s">{children}</Box>,
                );
            `;

            expect(code).toMatchSnapshot();
        });

        it('should keep the `styled` tag with long composition', () => {
            const code = transform({parser: 'babel'})`
                const Button = ({children, ...props}) =>
                styled(style1, style2, style3, style4, style5, style6)\`Box {margin-top: 10px;}
                \`(
                <Box {...props} size="s">{children}</Box>,
                );
            `;

            expect(code).toMatchSnapshot();
        });

        it('should keep process the callback and remove the comma dangle', () => {
            const code = transform({parser: 'babel'})`
            const Button = components.map(x => styled\`
                div {color: green;}
            \`(
                <div {...x.props} size="m">
                    {x.map(child => styled\`
                        child {opacity: .5}
                    \`(
                        <child>{child}</child>
                    ))}
                </div>,
            ));
            `;

            expect(code).toMatchSnapshot();
        });
    });
});
