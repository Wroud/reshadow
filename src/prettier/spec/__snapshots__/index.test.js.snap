// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prettier css should keep the capitalized tag 1`] = `
"Button {
  color: red;
}

Button[disabled] {
  color: red;
}

Container Button content {
  font-size: 10px;
}
"
`;

exports[`prettier js should keep process the callback and remove the comma dangle 1`] = `
"const Button = components.map(x => styled\`
  div {
    color: green;
  }
\`(
  <div {...x.props} size=\\"m\\">
    {x.map(child => styled\`
      child {
        opacity: 0.5;
      }
    \`(<child>{child}</child>))}
  </div>
));
"
`;

exports[`prettier js should keep the \`styled\` tag in arrow function and transform the css 1`] = `
"const Button = ({ children, ...props }) => styled\`
  Box {
    margin-top: 10px;
  }
\`(
  <Box {...props} size=\\"s\\">
    {children}
  </Box>
);
"
`;

exports[`prettier js should keep the \`styled\` tag with composition 1`] = `
"const Button = ({ children, ...props }) => styled(style1, style2)\`
  Box {
    margin-top: 10px;
  }
  button[disabled] {
    color: gray;
  }
\`(
  <Box {...props} size=\\"s\\">
    {children}
  </Box>
);
"
`;

exports[`prettier js should keep the \`styled\` tag with long composition 1`] = `
"const Button = ({ children, ...props }) => styled(
  style1,
  style2,
  style3,
  style4,
  style5,
  style6
)\`
  Box {
    margin-top: 10px;
  }
\`(
  <Box {...props} size=\\"s\\">
    {children}
  </Box>
);
"
`;
